From 2a7f5ce727796c67db009815aec76523367fac12 Mon Sep 17 00:00:00 2001
From: Hannes Winkler <hanneswinkler2000@web.de>
Date: Thu, 9 Nov 2023 15:04:30 +0100
Subject: [PATCH 3/3] support cross-building linux engine from macos

build/config/BUILDCONFIG.gn:
Use mac toolchain when building for linux

build/config/sysroot.gni:
Select sysroot based on target_cpu and allow arm sysroots

build/toolchain/linux/install-sysroot.py:
Allow downloading linux sysroots on mac

build/toolchain/linux/BUILD.gn:
Use //buildtools/mac-x64/clang... when running on mac
---
 build/config/BUILDCONFIG.gn                    | 14 +++++++++-----
 build/config/sysroot.gni                       | 12 ++++++------
 build/linux/sysroot_scripts/install-sysroot.py |  2 +-
 build/toolchain/linux/BUILD.gn                 | 14 +++++++++-----
 4 files changed, 25 insertions(+), 17 deletions(-)

diff --git a/build/config/BUILDCONFIG.gn b/build/config/BUILDCONFIG.gn
index 323a6ad..5db1aca 100644
--- a/build/config/BUILDCONFIG.gn
+++ b/build/config/BUILDCONFIG.gn
@@ -558,16 +558,20 @@ if (custom_toolchain != "") {
     set_default_toolchain("//build/toolchain/android:$current_cpu")
   }
 } else if (is_linux) {
+  if (host_os == "linux") {
+    if (is_clang) {
+      host_toolchain = "//build/toolchain/linux:clang_$host_cpu"
+    } else {
+      host_toolchain = "//build/toolchain/linux:$host_cpu"
+    }
+  } else if (host_os == "mac") {
+    host_toolchain = "//build/toolchain/mac:clang_$host_cpu"
+  }
   if (is_clang) {
-    host_toolchain = "//build/toolchain/linux:clang_$host_cpu"
     set_default_toolchain("//build/toolchain/linux:clang_$current_cpu")
   } else {
-    host_toolchain = "//build/toolchain/linux:$host_cpu"
     set_default_toolchain("//build/toolchain/linux:$current_cpu")
   }
-  if (is_chromeos && cros_use_custom_toolchain) {
-    set_default_toolchain("//build/toolchain/cros:target")
-  }
 } else if (is_mac) {
   host_toolchain = "//build/toolchain/mac:clang_$host_cpu"
   set_default_toolchain("//build/toolchain/mac:clang_$current_cpu")
diff --git a/build/config/sysroot.gni b/build/config/sysroot.gni
index 2c225a0..5361d8d 100644
--- a/build/config/sysroot.gni
+++ b/build/config/sysroot.gni
@@ -24,19 +24,19 @@ if (current_toolchain == default_toolchain && target_sysroot != "") {
   sysroot = rebase_path("$llvm_android_toolchain_root/sysroot")
 } else if (is_linux && !is_chromeos) {
   if (use_default_linux_sysroot && !is_fuchsia) {
-    if (current_cpu == "x64") {
+    if (target_cpu == "x64") {
       sysroot = rebase_path("//build/linux/debian_sid_amd64-sysroot")
-    } else if (current_cpu == "arm64") {
+    } else if (target_cpu == "arm64") {
       sysroot = rebase_path("//build/linux/debian_sid_arm64-sysroot")
-    } else if (current_cpu == "arm") {
+    } else if (target_cpu == "arm") {
       sysroot = rebase_path("//build/linux/debian_sid_arm-sysroot")
     } else {
-      assert(false, "Unknown CPU: $current_cpu.")
+      assert(false, "Unsupported cpu: $target_cpu")
     }
-
+    
     assert(
         exec_script("//build/dir_exists.py", [ sysroot ], "string") == "True",
-        "Missing sysroot ($sysroot). To fix, run: build/linux/sysroot_scripts/install-sysroot.py --arch=$current_cpu")
+        "Missing sysroot ($sysroot). To fix, run: build/linux/sysroot_scripts/install-sysroot.py --arch=$target_cpu")
   } else {
     sysroot = ""
   }
diff --git a/build/linux/sysroot_scripts/install-sysroot.py b/build/linux/sysroot_scripts/install-sysroot.py
index d3e7664..6dfbe42 100755
--- a/build/linux/sysroot_scripts/install-sysroot.py
+++ b/build/linux/sysroot_scripts/install-sysroot.py
@@ -77,7 +77,7 @@ def main(args):
   parser.add_option('--print-hash',
                     help='Print the hash of the sysroot for the given arch.')
   options, _ = parser.parse_args(args)
-  if not sys.platform.startswith('linux'):
+  if not (sys.platform.startswith('linux') or sys.platform.startswith('darwin')):
     return 0
 
   if options.print_hash:
diff --git a/build/toolchain/linux/BUILD.gn b/build/toolchain/linux/BUILD.gn
index 05630d0..2e89769 100644
--- a/build/toolchain/linux/BUILD.gn
+++ b/build/toolchain/linux/BUILD.gn
@@ -21,14 +21,18 @@ if (use_goma) {
   compiler_prefix = ""
 }
 
-if (host_cpu == "arm64") {
-  rebased_clang_dir =
-      rebase_path("//buildtools/linux-arm64/clang/bin", root_build_dir)
+if (host_os == "linux" && host_cpu == "x64") {
+  host_dir = "linux-x64"
+} else if (host_os == "linux" && host_cpu == "arm64") {
+  host_dir = "linux-arm64"
+} else if (host_os == "mac" && host_cpu == "x64") {
+  host_dir = "mac-x64"
 } else {
-  rebased_clang_dir =
-      rebase_path("//buildtools/linux-x64/clang/bin", root_build_dir)
+  assert(false, "Unsupported host OS & CPU combination: $host_os, $host_cpu")
 }
 
+rebased_clang_dir = rebase_path("//buildtools/$host_dir/clang/bin", root_build_dir)
+
 gcc_toolchain("arm") {
   prefix = "arm-linux-gnueabihf-"
   if (toolchain_prefix != "") {
-- 
2.32.1 (Apple Git-133)

